source/common/mongo/proxy.cc

```diff
@@ -241,5 +245,54 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
   return DecoderPtr{new DecoderImpl(callbacks)};
 }
 
+FaultConfigImpl::FaultConfigImpl(const Json::Object& fault_config)
+    : delay_percent_(
+          static_cast<uint32_t>(fault_config.getObject("fixed_delay")->getInteger("percent"))),
+      duration_ms_(
+          static_cast<uint64_t>(fault_config.getObject("fixed_delay")->getInteger("duration_ms"))) {
+}
+
+Optional<uint64_t> ProxyFilter::delayDuration() {
+  Optional<uint64_t> result;
+
+  if (!fault_config_)
+    return result;
+
+  if (!runtime_.snapshot().featureEnabled("mongo.fault.delay.percent",
+                                          fault_config_->delayPercent())) {
+    return result;
+  }
+
+  uint64_t duration = runtime_.snapshot().getInteger("mongo.fault.delay.duration_ms",
+                                                     fault_config_->delayDuration());
+
+  // Delay only if the duration is > 0ms.
+  if (duration > 0) {
+    result.value(duration);
+  }
+
+  return result;
+}
+
+void ProxyFilter::delayInjectionTimerCallback() {
```

Comment:
Active query list is used just to match request/response in order to capture timing and other full request processing stats.

When new operation coming in `Query` for example, we recognize that it's a `Query` operation and we delay all of the further processing (if needed). The active query will be processed once we get a reply for that (based on id).

I do not see any better way of handling this but open for any suggestions.